src/Features/Core/Portable/InlineMethod/AbstractInlineMethodRefactoringProvider.InlineContext.cs

```diff
@@ -0,0 +1,647 @@
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.InlineMethod
+{
+    internal abstract partial class AbstractInlineMethodRefactoringProvider<TMethodDeclarationSyntax, TStatementSyntax, TExpressionSyntax, TInvocationSyntax>
+    {
+        private readonly struct InlineMethodContext
+        {
+            /// <summary>
+            /// Statements that should be inserted before the <see cref="StatementContainingCallee"/>.
+            /// </summary>
+            public ImmutableArray<TStatementSyntax> StatementsToInsertBeforeCallee { get; }
+
+            /// <summary>
+            /// Statement containing the callee. All the generated declarations should be put before this node.
+            /// Example:
+            /// void Caller()
+            /// {
+            ///     var x = Callee(Foo());
+            /// }
+            /// void Callee(int i) { DoSomething(); }
+            /// LocalDeclarationSyntaxNode of x will the <see cref="StatementContainingCallee"/>.
+            /// And if there is any statements needs inserted, it needs to be inserted before this node.
+            /// </summary>
+            public TStatementSyntax StatementContainingCallee { get; }
+
+            /// <summary>
+            /// Inline content for the callee method. It should replace <see cref="SyntaxNodeToReplace"/>.
+            /// </summary>
+            public SyntaxNode InlineSyntaxNode { get; }
+
+            /// <summary>
+            /// SyntaxNode needs to be replaced by <see cref="InlineSyntaxNode"/>
+            /// </summary>
+            public SyntaxNode SyntaxNodeToReplace { get; }
+
+            /// <summary>
+            /// Indicate if <see cref="InlineSyntaxNode"/> has AwaitExpression in it.
+            /// </summary>
+            public bool ContainsAwaitExpression { get; }
+
+            public InlineMethodContext(
+                ImmutableArray<TStatementSyntax> statementsToInsertBeforeCallee,
+                TStatementSyntax statementContainingCallee,
+                SyntaxNode inlineSyntaxNode,
+                SyntaxNode syntaxNodeToReplace,
+                bool containsAwaitExpression)
+            {
+                StatementsToInsertBeforeCallee = statementsToInsertBeforeCallee;
+                StatementContainingCallee = statementContainingCallee;
+                InlineSyntaxNode = inlineSyntaxNode;
+                SyntaxNodeToReplace = syntaxNodeToReplace;
+                ContainsAwaitExpression = containsAwaitExpression;
+            }
+        }
+
+        private async Task<InlineMethodContext> GetInlineMethodContextAsync(
+            Document document,
+            TMethodDeclarationSyntax calleeMethodNode,
+            TInvocationSyntax calleeInvocationNode,
+            IMethodSymbol calleeMethodSymbol,
+            TExpressionSyntax rawInlineExpression,
+            TStatementSyntax statementContainsCallee,
```

Comment:
nit: you don't have to change thigns, but i consider this to be a relatively mishmashed disorganized set of of parameters.  i.e. it would make sense to group a bunch f related things together.  like "calleeMethodNode, calleeStatementNode, calleeInvocationNode`.  Nodes like the `rawInlineExpression` should not be intermixed with that imo.

Reply:
oh, i may be misunderstanding this in terms of what is being refered to by what here.

Reply:
(i.e. my brain gets confused with caller/callee).